server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java

```diff
@@ -0,0 +1,362 @@
+
+        // This implementation skips logging Deletes if the operation is conditional (thus no
+        // Resource ID present)
+
+        auditEventList =
+            createAuditEventCore(
+                resources, BalpProfileEnum.PATIENT_DELETE, BalpProfileEnum.BASIC_DELETE);
+        break;
+
+      default:
+        break;
+    }
+
+    // TODO Investigate bulk saving (batch processing) instead to improve performance. We'll
+    // probably need a mechanism for chunking e.g. 100, 200, or 500 batch
+    for (AuditEvent auditEvent : auditEventList) {
+      auditEvent.getPeriod().setEnd(new Date());
+      try {
+        this.httpFhirClient.postResource(auditEvent);
+      } catch (IOException exception) {
+        ExceptionUtil.throwRuntimeExceptionAndLog(logger, exception.getMessage(), exception);
+      }
+    }
+  }
+
+  private String getResourceTemplate(String resourceName, String resourceId) {
+    return String.format("{ \"resourceType\": \"%s\", \"id\": \"%s\"}", resourceName, resourceId);
+  }
+
+  private List<AuditEvent> createAuditEventCore(
+      List<IBaseResource> resources, BalpProfileEnum patientProfile, BalpProfileEnum basicProfile) {
+    List<AuditEvent> auditEventList = new ArrayList<>();
+
+    for (IBaseResource iBaseResource : resources) {
+      try {
+        Preconditions.checkState(iBaseResource instanceof DomainResource);
+
+        DomainResource resource = (DomainResource) iBaseResource;
+
+        Set<String> patientIds = patientFinder.findPatientIds(resource);
+
+        if (!patientIds.isEmpty()) {
+          auditEventList.add(createAuditEvent(resource, patientProfile, patientIds));
+        } else {
+          auditEventList.add(createAuditEvent(resource, basicProfile, Set.of()));
+        }
+      } catch (IllegalStateException exception) {
+        logger.error(exception.getMessage(), exception);
+      }
+    }
+
+    return auditEventList;
+  }
+
+  @Nullable
+  private AuditEvent.AuditEventOutcome mapOutcomeErrorCode(
+      OperationOutcome.IssueSeverity issueSeverity) {
+    AuditEvent.AuditEventOutcome errorCode = null;
+    if (OperationOutcome.IssueSeverity.FATAL.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._12;
+    } else if (OperationOutcome.IssueSeverity.ERROR.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._8;
+    } else if (OperationOutcome.IssueSeverity.WARNING.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._4;
+    } else if (OperationOutcome.IssueSeverity.INFORMATION.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._0;
+    }
+    return errorCode;
+  }
+
+  private List<IBaseResource> extractFhirResources(String serverContentResponse) {
+    List<IBaseResource> resourceList = new ArrayList<>();
+
+    // This condition handles operations with no response body returned e.g. POST/PUT requests with
+    // Prefer: return=minimal HTTP header
```

Comment:
In this case, can't we still access the full resource by looking at the content of the `requestDetailsReader`? An important issue that I see here is that a patient related resource might be added or updated without any reference to that patient.

Reply:
If you feel this is a good/doable idea but it is a lot of work, feel free to add a TODO and leave this for a future follow-up PR.

Reply:
Yeah there is this ticket I created earlier, please review and update anything I missed - https://github.com/google/fhir-gateway/issues/411